### PR TITLE
MI additions (with correct commit messages)

### PR DESCRIPTION
**Pull request by ethanperry1**

### CODE REVIEW

**Comment by ethanperry1** [comment]
There was an issue with commit messages on this PR, which was causing the CI to fail (https://github.com/princjef/irontask/pull/59/files). I addressed your PR comment and rebased the commit message wording. 

I made the identity a peer dependency, made the aad auth optional, and fixed the promise issue you mentioned as well.

File: src/client/client.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@
  */
 
 import { SqlQuerySpec } from '@azure/cosmos';
+import { TokenCredential } from '@azure/identity';
```
Comment (princjef) [inline src/client/client.ts:7]: nit: use type import where you pull in `@azure/identity` to ensure you're not depending on the code
```suggestion
import type { TokenCredential } from '@azure/identity';
```

**Review by princjef — APPROVED** [review]

**Comment by ethanperry1** [comment]
Test run for MI locally (using az cli credential):


> irontask@1.5.0 pretest
> npm run build:verify


> irontask@1.5.0 prebuild:verify
> npm run clean && npm run format:verify && npm run lint:verify


> irontask@1.5.0 clean
> rimraf dist


> irontask@1.5.0 format:verify
> prettier --check --loglevel warn "src/**/*.ts" "*.md"


> irontask@1.5.0 lint:verify
> eslint -c .eslintrc.js --ext .ts src


Could not find implementations for the following rules specified in the configuration:
    naming-convention
    no-else-after-return
    no-function-constructor-with-string-args
    no-increment-decrement
    no-typeof-undefined
    no-unnecessary-else
    no-unnecessary-override
    no-unused
    no-var-before-return
    promise-must-complete
    ter-prefer-arrow-callback
Try upgrading TSLint and/or ensuring that you have all necessary custom rules installed.
If TSLint was recently upgraded, you may have old rules configured which need to be cleaned up.
        

> irontask@1.5.0 build:verify
> tsc


> irontask@1.5.0 test
> jest --testTimeout=60000 --runInBand

PASS src/test/client.spec.ts (76.318 s)
PASS src/test/listener.spec.ts (62.817 s)
PASS src/test/task.spec.ts (41.055 s)
PASS src/test/readonlyTask.spec.ts (33.05 s)
PASS src/query/serializer/expression.spec.ts
PASS src/query/serializer/projection.spec.ts
PASS src/utils/computeNextRun.spec.ts
PASS src/query/serializer/query.spec.ts
PASS src/utils/batchIterator.spec.ts
PASS src/task/status.spec.ts
PASS src/utils/retry.spec.ts
PASS src/utils/fuzz.spec.ts
PASS src/utils/isValidNextRun.spec.ts
----------------------|---------|----------|---------|---------|-----------------------------------------------------------------------
File                  | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                                                     
----------------------|---------|----------|---------|---------|-----------------------------------------------------------------------
All files             |   94.94 |    91.37 |    96.2 |   94.67 |                                                                       
 src                  |     100 |      100 |   58.33 |     100 |                                                                       
  constants.ts        |     100 |      100 |     100 |     100 |                                                                       
  index.ts            |     100 |      100 |   58.33 |     100 |                                                                       
 src/client           |   86.65 |    77.86 |   93.71 |   86.07 |                                                                       
  client.ts           |   93.08 |    87.71 |   98.48 |   92.48 | 58,172-186,1074-1095                                                  
  cosmos.ts           |   70.42 |    71.15 |      86 |   70.28 | 66-74,162,244-284,334-340,407,444,478,569-576,581-585,590-606,615,623 
  index.ts            |     100 |      100 |     100 |     100 |                                                                       
  listener.ts         |   94.17 |    81.25 |   92.85 |   93.87 | 214-231,338,421                                                       
  scoped.ts           |     100 |    33.33 |     100 |     100 | 87-133,183                                                            
 src/error            |    97.5 |    83.33 |     100 |    97.5 |                                                                       
  codes.ts            |     100 |      100 |     100 |     100 |                                                                       
  error.ts            |    90.9 |       75 |     100 |    90.9 | 30                                                                    
  index.ts            |     100 |      100 |     100 |     100 |                                                                       
  messages.ts         |     100 |      100 |     100 |     100 |                                                                       
 src/interceptor      |     100 |      100 |     100 |     100 |                                                                       
  index.ts            |     100 |      100 |     100 |     100 |                                                                       
  processor.ts        |     100 |      100 |     100 |     100 |                                                                       
  schema.ts           |     100 |      100 |     100 |     100 |                                                                       
 src/query            |     100 |      100 |      40 |     100 |                                                                       
  index.ts            |     100 |      100 |      40 |     100 |                                                                       
 src/query/builder    |     100 |      100 |     100 |     100 |                                                                       
  index.ts            |     100 |      100 |     100 |     100 |                                                                       
  property.ts         |     100 |      100 |     100 |     100 |                                                                       
  query.ts            |     100 |      100 |     100 |     100 |                                                                       
 src/query/serializer |    99.3 |      100 |     100 |    99.3 |                                                                       
  expression.ts       |     100 |      100 |     100 |     100 |                                                                       
  projection.ts       |   97.14 |      100 |     100 |   97.14 | 103                                                                   
  query.ts            |     100 |      100 |     100 |     100 |                                                                       
 src/query/symbols    |     100 |      100 |     100 |     100 |                                                                       
  functions.ts        |     100 |      100 |     100 |     100 |                                                                       
  index.ts            |     100 |      100 |     100 |     100 |                                                                       
  operators.ts        |     100 |      100 |     100 |     100 |                                                                       
  properties.ts       |     100 |      100 |     100 |     100 |                                                                       
 src/task             |   95.71 |     88.4 |   99.43 |   95.62 |                                                                       
  active.ts           |   93.82 |       90 |     100 |   93.78 | 460-463,513-514,598,610-618,643                                       
  data.ts             |   98.92 |    90.69 |     100 |   98.92 | 464                                                                   
  index.ts            |     100 |      100 |     100 |     100 |                                                                       
  properties.ts       |   80.64 |    33.33 |     100 |   80.64 | 207-250                                                               
  query.ts            |     100 |      100 |     100 |     100 |                                                                       
  readonly.ts         |   96.87 |      100 |   96.29 |   96.66 | 98                                                                    
  status.ts           |     100 |      100 |     100 |     100 |                                                                       
  task.ts             |     100 |      100 |     100 |     100 |                                                                       
  types.ts            |     100 |      100 |     100 |     100 |                                                                       
 src/utils            |      98 |    93.75 |     100 |   97.82 |                                                                       
  batchIterator.ts    |   92.85 |    66.66 |     100 |    90.9 | 8                                                                     
  computeNextRun.ts   |     100 |      100 |     100 |     100 |                                                                       
  delay.ts            |     100 |      100 |     100 |     100 |                                                                       
  fuzz.ts             |     100 |      100 |     100 |     100 |                                                                       
  isValidNextRun.ts   |     100 |      100 |     100 |     100 |                                                                       
  retry.ts            |     100 |      100 |     100 |     100 |                                                                       
----------------------|---------|----------|---------|---------|-----------------------------------------------------------------------

Test Suites: 13 passed, 13 total
Tests:       352 passed, 352 total
Snapshots:   0 total
Time:        215.261 s, estimated 221 s
Ran all test suites.

Test run for key based auth:

> irontask@1.5.0 pretest
> npm run build:verify


> irontask@1.5.0 prebuild:verify
> npm run clean && npm run format:verify && npm run lint:verify


> irontask@1.5.0 clean
> rimraf dist


> irontask@1.5.0 format:verify
> prettier --check --loglevel warn "src/**/*.ts" "*.md"


> irontask@1.5.0 lint:verify
> eslint -c .eslintrc.js --ext .ts src


Could not find implementations for the following rules specified in the configuration:
    naming-convention
    no-else-after-return
    no-function-constructor-with-string-args
    no-increment-decrement
    no-typeof-undefined
    no-unnecessary-else
    no-unnecessary-override
    no-unused
    no-var-before-return
    promise-must-complete
    ter-prefer-arrow-callback
Try upgrading TSLint and/or ensuring that you have all necessary custom rules installed.
If TSLint was recently upgraded, you may have old rules configured which need to be cleaned up.
        

> irontask@1.5.0 build:verify
> tsc


> irontask@1.5.0 test
> jest --testTimeout=60000 --runInBand

PASS src/test/client.spec.ts (49.214 s)
PASS src/test/listener.spec.ts (38.534 s)
PASS src/test/task.spec.ts (14.499 s)
PASS src/test/readonlyTask.spec.ts (8.334 s)
PASS src/query/serializer/expression.spec.ts
PASS src/query/serializer/projection.spec.ts
PASS src/utils/computeNextRun.spec.ts
PASS src/query/serializer/query.spec.ts
PASS src/task/status.spec.ts
PASS src/utils/fuzz.spec.ts
PASS src/utils/retry.spec.ts
PASS src/utils/batchIterator.spec.ts
PASS src/utils/isValidNextRun.spec.ts
----------------------|---------|----------|---------|---------|--------------------------------------------------------------------
File                  | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                                                  
----------------------|---------|----------|---------|---------|--------------------------------------------------------------------
All files             |   94.82 |    91.37 |    95.6 |   94.47 |                                                                    
 src                  |     100 |      100 |   58.33 |     100 |                                                                    
  constants.ts        |     100 |      100 |     100 |     100 |                                                                    
  index.ts            |     100 |      100 |   58.33 |     100 |                                                                    
 src/client           |   86.21 |    77.86 |   91.82 |   85.38 |                                                                    
  client.ts           |   94.68 |    87.71 |   98.48 |   94.21 | 58,131-148,1071,1083-1095                                          
  cosmos.ts           |    66.9 |    71.15 |      80 |   65.94 | 95-185,207-244,284,334-340,407,444,569-576,581-585,590-606,615,623 
  index.ts            |     100 |      100 |     100 |     100 |                                                                    
  listener.ts         |   94.17 |    81.25 |   92.85 |   93.87 | 214-231,338,421                                                    
  scoped.ts           |     100 |    33.33 |     100 |     100 | 87-133,183                                                         
 src/error            |    97.5 |    83.33 |     100 |    97.5 |                                                                    
  codes.ts            |     100 |      100 |     100 |     100 |                                                                    
  error.ts            |    90.9 |       75 |     100 |    90.9 | 30                                                                 
  index.ts            |     100 |      100 |     100 |     100 |                                                                    
  messages.ts         |     100 |      100 |     100 |     100 |                                                                    
 src/interceptor      |     100 |      100 |     100 |     100 |                                                                    
  index.ts            |     100 |      100 |     100 |     100 |                                                                    
  processor.ts        |     100 |      100 |     100 |     100 |                                                                    
  schema.ts           |     100 |      100 |     100 |     100 |                                                                    
 src/query            |     100 |      100 |      40 |     100 |                                                                    
  index.ts            |     100 |      100 |      40 |     100 |                                                                    
 src/query/builder    |     100 |      100 |     100 |     100 |                                                                    
  index.ts            |     100 |      100 |     100 |     100 |                                                                    
  property.ts         |     100 |      100 |     100 |     100 |                                                                    
  query.ts            |     100 |      100 |     100 |     100 |                                                                    
 src/query/serializer |    99.3 |      100 |     100 |    99.3 |                                                                    
  expression.ts       |     100 |      100 |     100 |     100 |                                                                    
  projection.ts       |   97.14 |      100 |     100 |   97.14 | 103                                                                
  query.ts            |     100 |      100 |     100 |     100 |                                                                    
 src/query/symbols    |     100 |      100 |     100 |     100 |                                                                    
  functions.ts        |     100 |      100 |     100 |     100 |                                                                    
  index.ts            |     100 |      100 |     100 |     100 |                                                                    
  operators.ts        |     100 |      100 |     100 |     100 |                                                                    
  properties.ts       |     100 |      100 |     100 |     100 |                                                                    
 src/task             |   95.71 |     88.4 |   99.43 |   95.62 |                                                                    
  active.ts           |   93.82 |       90 |     100 |   93.78 | 460-463,513-514,598,610-618,643                                    
  data.ts             |   98.92 |    90.69 |     100 |   98.92 | 464                                                                
  index.ts            |     100 |      100 |     100 |     100 |                                                                    
  properties.ts       |   80.64 |    33.33 |     100 |   80.64 | 207-250                                                            
  query.ts            |     100 |      100 |     100 |     100 |                                                                    
  readonly.ts         |   96.87 |      100 |   96.29 |   96.66 | 98                                                                 
  status.ts           |     100 |      100 |     100 |     100 |                                                                    
  task.ts             |     100 |      100 |     100 |     100 |                                                                    
  types.ts            |     100 |      100 |     100 |     100 |                                                                    
 src/utils            |      98 |    93.75 |     100 |   97.82 |                                                                    
  batchIterator.ts    |   92.85 |    66.66 |     100 |    90.9 | 8                                                                  
  computeNextRun.ts   |     100 |      100 |     100 |     100 |                                                                    
  delay.ts            |     100 |      100 |     100 |     100 |                                                                    
  fuzz.ts             |     100 |      100 |     100 |     100 |                                                                    
  isValidNextRun.ts   |     100 |      100 |     100 |     100 |                                                                    
  retry.ts            |     100 |      100 |     100 |     100 |                                                                    
----------------------|---------|----------|---------|---------|--------------------------------------------------------------------

Test Suites: 13 passed, 13 total
Tests:       352 passed, 352 total
Snapshots:   0 total
Time:        112.308 s, estimated 215 s
Ran all test suites.

**Note that the MI tests took twice as long.** I have noticed in general that they take much longer, probably due to the resource polling involved in the arm requests.

File: src/client/client.ts [hunk 2]
```diff
@@ -106,24 +106,32 @@ export default class TaskClient {
...
+   * @param options             - Client creation options
    *
    * @returns Promise containing the initialized client
    *
    * @public
    */
   static async createFromCredential(
+    subscriptionId: string,
+    resourceGroupName: string,
+    accountEndpoint: string,
```
Comment (princjef) [inline src/client/client.ts:124]: what's the difference between `accountEndpoint` and `account`?

Comment (princjef) [reply src/client/client.ts:124]: ...and can they be combined into a single variable?

File: src/client/cosmos.ts [hunk 3]
```diff
@@ -45,63 +51,200 @@ const connectionPolicy: ConnectionPolicy = {
...
+  constructor(options: ManagementOptions) {
+    this._management = new CosmosDBManagementClient(
+      options.storageCredentials,
+      options.subscriptionId
+    );
+    this._options = options;
+  }
+
+  async createDatabaseIfNotExists(params: SqlDatabaseCreateUpdateParameters) {
+    // try {
```
Comment (princjef) [inline src/client/cosmos.ts:103]: delete?

File: src/client/cosmos.ts [hunk 4]
```diff
@@ -45,63 +51,200 @@ const connectionPolicy: ConnectionPolicy = {
...
+        this._options.resourceGroupName,
+        this._options.accountName,
+        this._options.databaseName,
+        params
+      )
+      .then(async poller => poller.pollUntilDone());
+  }
+
+  async createContainerIfNotExists(params: SqlContainerCreateUpdateParameters) {
+    // try {
```
Comment (princjef) [inline src/client/cosmos.ts:131]: delete?

File: src/client/cosmos.ts [hunk 5]
```diff
@@ -45,63 +51,200 @@ const connectionPolicy: ConnectionPolicy = {
...
+        databaseName: database,
+        containerName: collection
+      });
+
+      await managementClient.createDatabaseIfNotExists({
+        resource: {
           id: database
-        },
-        {
-          offerThroughput: DEFAULT_DATABASE_THROUGHPUT
```
Comment (princjef) [inline src/client/cosmos.ts:72]: why is this not in the new code?

File: src/client/cosmos.ts [hunk 6]
```diff
@@ -401,6 +561,8 @@ export default class CosmosDbClient {
     // We scrub the error first for sensitive information
     const scrubbedError = CosmosDbClient._scrubError(err);
 
+    // console.error(err);
```
Comment (princjef) [inline src/client/cosmos.ts:564]: delete

File: src/client/client.ts (hunk 2 above)
Comment (ethanperry1) [reply src/client/client.ts:124]: I thought they needed to be separate, upon further inspection I can likely format this based on the account name

File: src/client/cosmos.ts [hunk 7]
```diff
@@ -655,6 +636,10 @@ export default class CosmosDbClient {
       (err && retryableNetworkingErrors.includes(err.code))
     );
   }
+
+  private static _formatEndpoint(accountName: string): string {
+    return `https://${accountName}.documents.azure.com:443/`;
```
Comment (princjef) [inline src/client/cosmos.ts:641]: ask for the endpoint, not the name. the `documents.azure.com` suffix only applies to public cloud so it's important that we allow people to set the endpoint that is relevant to them

**Review by princjef — APPROVED** [review]
just flip `accountName` to `accountEndpoint`

**Pull request merged**